wasm build fails since oxc_resolver 11.5.1

When compiling `oxc_resolver` for wasm after `11.5.1`, I get the following error:

```
   Compiling oxc_resolver v11.8.2
error[E0599]: no method named `to_file_path` found for struct `Url` in the current scope
   --> /Users/ives/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/oxc_resolver-11.8.2/src/lib.rs:375:18
    |
373 |               let path = Url::parse(specifier)
    |  ________________________-
374 | |                 .map_err(|_| unsupported_error.clone())?
375 | |                 .to_file_path()
    | |                 -^^^^^^^^^^^^ method not found in `Url`
    | |_________________|
```

I think that's because in this commit (https://github.com/oxc-project/oxc-resolver/commit/195e2f9357fd6378dbf78a5b51b33a914407d039) `Url::parse(...).to_file_path()` was added, and that is not supported in wasm. I wonder if it would be sufficient to add a pragma to this block:

```
        if specifier.starts_with("file://") {
            let unsupported_error = ResolveError::PathNotSupported(specifier.into());

            let path = Url::parse(specifier)
                .map_err(|_| unsupported_error.clone())?
                .to_file_path()
                .map_err(|()| unsupported_error)?;

            let owned = path.to_string_lossy().into_owned();
            specifier_owned = Some(owned);
            specifier = specifier_owned.as_deref().unwrap();
        }
```

That ensures we don't compile that block for wasm targets?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -8,15 +8,13 @@ on:
     types:
       - opened
       - synchronize
-    paths-ignore:
+    paths-ignore: &paths-ignore
       - "**/*.md"
       - "!.github/workflows/ci.yml"
   push:
     branches:
       - main
-    paths-ignore:
-      - "**/*.md"
-      - "!.github/workflows/ci.yml"
+    paths-ignore: *paths-ignore
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
@@ -62,21 +60,37 @@ jobs:
         with:
           files: .
 
-  wasm:
-    name: Check Wasm
+  wasm32-wasip1:
+    name: Check wasm32-wasip1
     runs-on: ubuntu-latest
     steps:
       - uses: taiki-e/checkout-action@b13d20b7cda4e2f325ef19895128f7ff735c0b3d # v1.3.1
 
       - uses: oxc-project/setup-rust@f78b3964e121f889426634f0eaeeb09fccecac08 # v1.0.6
         with:
-          cache-key: wasm
+          cache-key: wasm32-wasip1
           save-cache: ${{ github.ref_name == 'main' }}
 
       - name: Check
         run: |
           rustup target add wasm32-wasip1
           cargo check --all-features --target wasm32-wasip1
+
+  wasm32-unknown-unknown:
+    name: Check wasm32-unknown-unknown
+    runs-on: ubuntu-latest
+    steps:
+      - uses: taiki-e/checkout-action@b13d20b7cda4e2f325ef19895128f7ff735c0b3d # v1.3.1
+
+      - uses: oxc-project/setup-rust@f78b3964e121f889426634f0eaeeb09fccecac08 # v1.0.6
+        with:
+          cache-key: wasm32-unknown-unknown
+          save-cache: ${{ github.ref_name == 'main' }}
+
+      - name: Check
+        run: |
+          rustup target add wasm32-unknown-unknown
+          cargo check --all-features --target wasm32-unknown-unknown
 
   wasi:
     name: Test wasi target

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -88,7 +88,6 @@ serde_json = { version = "1", features = ["preserve_order"] } # preserve_order: 
 simdutf8 = { version = "0.1" }
 thiserror = "2"
 tracing = "0.1"
-url = "2"
 
 pnp = { version = "0.12.2", optional = true }
 
@@ -99,6 +98,7 @@ libc = "0.2"
 
 [target.'cfg(target_os = "windows")'.dependencies]
 windows = { version = "0.62.0", features = ["Win32_Storage_FileSystem"] }
+url = "2"
 
 [dev-dependencies]
 criterion2 = { version = "3.0.2", default-features = false }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -359,7 +359,12 @@ impl<Fs: FileSystem> ResolverGeneric<Fs> {
             return Ok(path);
         }
 
-        let specifier = resolve_file_protocol(specifier);
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "windows")] {
+                let specifier = resolve_file_protocol(specifier)?;
+                let specifier = specifier.as_ref();
+            }
+        };
 
         let result = match Path::new(&specifier).components().next() {
             // 2. If X begins with '/'
@@ -2088,19 +2093,14 @@ impl<Fs: FileSystem> ResolverGeneric<Fs> {
 }
 
 #[cfg(target_os = "windows")]
-fn resolve_file_protocol(specifier: &str) -> Cow<'_, str> {
+fn resolve_file_protocol(specifier: &str) -> Result<Cow<'_, str>, ResolveError> {
     if specifier.starts_with("file://") {
         url::Url::parse(&specifier)
             .map_err(|_| ())
             .and_then(|p| p.to_file_path())
             .map(|path| Cow::Owned(path.to_string_lossy().to_string()))
-            .map_err(|_| ResolveError::PathNotSupported(PathBuf::from(specifier.as_ref())))?;
+            .map_err(|_| ResolveError::PathNotSupported(PathBuf::from(specifier)))
     } else {
-        Cow::Borrowed(specifier)
+        Ok(Cow::Borrowed(specifier))
     }
-}
-
-#[cfg(not(target_os = "windows"))]
-fn resolve_file_protocol(specifier: &str) -> &str {
-    specifier
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -65,17 +65,6 @@ mod windows;
 #[cfg(test)]
 mod tests;
 
-use rustc_hash::FxHashSet;
-use std::{
-    borrow::Cow,
-    cmp::Ordering,
-    ffi::OsStr,
-    fmt, iter,
-    path::{Component, Path, PathBuf},
-    sync::Arc,
-};
-use url::Url;
-
 pub use crate::{
     builtins::NODEJS_BUILTINS,
     cache::{Cache, CachedPath},
@@ -98,6 +87,15 @@ pub use crate::{
 use crate::{
     context::ResolveContext as Ctx, path::SLASH_START, specifier::Specifier,
     tsconfig_context::TsconfigResolveContext,
+};
+use rustc_hash::FxHashSet;
+use std::{
+    borrow::Cow,
+    cmp::Ordering,
+    ffi::OsStr,
+    fmt, iter,
+    path::{Component, Path, PathBuf},
+    sync::Arc,
 };
 
 type ResolveResult = Result<Option<CachedPath>, ResolveError>;
@@ -361,24 +359,9 @@ impl<Fs: FileSystem> ResolverGeneric<Fs> {
             return Ok(path);
         }
 
-        #[allow(unused_assignments)]
-        let mut specifier_owned: Option<String> = None;
-        let mut specifier = specifier;
+        let specifier = resolve_file_protocol(specifier);
 
-        if specifier.starts_with("file://") {
-            let unsupported_error = ResolveError::PathNotSupported(specifier.into());
-
-            let path = Url::parse(specifier)
-                .map_err(|_| unsupported_error.clone())?
-                .to_file_path()
-                .map_err(|()| unsupported_error)?;
-
-            let owned = path.to_string_lossy().into_owned();
-            specifier_owned = Some(owned);
-            specifier = specifier_owned.as_deref().unwrap();
-        }
-
-        let result = match Path::new(specifier).components().next() {
+        let result = match Path::new(&specifier).components().next() {
             // 2. If X begins with '/'
             Some(Component::RootDir | Component::Prefix(_)) => {
                 self.require_absolute(cached_path, specifier, ctx)
@@ -2102,4 +2085,22 @@ impl<Fs: FileSystem> ResolverGeneric<Fs> {
             _ => Ok(None),
         }
     }
+}
+
+#[cfg(target_os = "windows")]
+fn resolve_file_protocol(specifier: &str) -> Cow<'_, str> {
+    if specifier.starts_with("file://") {
+        url::Url::parse(&specifier)
+            .map_err(|_| ())
+            .and_then(|p| p.to_file_path())
+            .map(|path| Cow::Owned(path.to_string_lossy().to_string()))
+            .map_err(|_| ResolveError::PathNotSupported(PathBuf::from(specifier.as_ref())))?;
+    } else {
+        Cow::Borrowed(specifier)
+    }
+}
+
+#[cfg(not(target_os = "windows"))]
+fn resolve_file_protocol(specifier: &str) -> &str {
+    specifier
 }

--- a/src/tests/resolve.rs
+++ b/src/tests/resolve.rs
@@ -1,7 +1,5 @@
 //! <https://github.com/webpack/enhanced-resolve/blob/main/test/resolve.test.js>
 
-use url::Url;
-
 use crate::{Resolution, ResolveError, ResolveOptions, Resolver};
 
 #[test]
@@ -12,12 +10,9 @@ fn resolve() {
 
     let main1_js_path = f.join("main1.js").to_string_lossy().to_string();
 
-    let file_protocol_path = Url::from_file_path(main1_js_path.clone()).unwrap();
-
     #[rustfmt::skip]
     let pass = [
         ("absolute path", f.clone(), main1_js_path.as_str(), f.join("main1.js")),
-        ("file protocol absolute path", f.clone(), file_protocol_path.as_str(), f.join("main1.js")),
         ("file with .js", f.clone(), "./main1.js", f.join("main1.js")),
         ("file without extension", f.clone(), "./main1", f.join("main1.js")),
         ("another file with .js", f.clone(), "./a.js", f.join("a.js")),
@@ -67,13 +62,6 @@ fn resolve() {
         }
         assert_eq!(resolved_path, Some(expected), "{comment} {path:?} {request}");
     }
-
-    #[cfg(windows)]
-    let resolve_error = ResolveError::NotFound("\\\\.\\main.js".into());
-    #[cfg(not(windows))]
-    let resolve_error = ResolveError::PathNotSupported("file://./main.js".into());
-
-    assert_eq!(resolver.resolve(f, "file://./main.js"), Err(resolve_error));
 }
 
 #[test]
@@ -273,4 +261,25 @@ fn resolve_normalized_on_windows() {
         resolution.map(|r| r.to_string_lossy().into_owned()),
         Ok(absolute_str.clone().into_owned())
     );
+}
+
+#[cfg(windows)]
+#[test]
+fn file_protocol() {
+    use url::Url;
+
+    let f = super::fixture();
+
+    let main1_js_path = f.join("main1.js").to_string_lossy().to_string();
+    let file_protocol_path = Url::from_file_path(main1_js_path.clone()).unwrap();
+
+    let resolver = Resolver::default();
+
+    let resolution = resolver.resolve(&f, file_protocol_path.as_str()).ok();
+    let resolved_path = resolution.as_ref().map(Resolution::full_path);
+    assert_eq!(resolved_path, Some(f.join("main1.js")));
+
+    let resolve_error = ResolveError::NotFound("\\\\.\\main.js".into());
+
+    assert_eq!(resolver.resolve(f, "file://./main.js"), Err(resolve_error));
 }

